json_serializable/lib/src/type_helpers/iterable_helper.dart

```diff
@@ -70,14 +71,28 @@ class IterableHelper extends TypeHelper<TypeHelperContextWithConfig> {
         _coreSetChecker.isExactlyType(targetType))) {
       return null;
     }
+    final targetTypeIsNullable = targetType.isNullableType;
 
     final iterableGenericType = coreIterableGenericType(targetType);
 
     final itemSubVal = context.deserialize(iterableGenericType, closureArg);
 
-    var output = '$expression as List<dynamic>';
+    final useDynamic =
+        iterableGenericType.element.kind == ElementKind.TYPE_PARAMETER ||
+            iterableGenericType.element.kind == ElementKind.ENUM ||
+            !defaultProvided;
 
-    final targetTypeIsNullable = defaultProvided || targetType.isNullableType;
+    final isPrimitiveObject = iterableGenericType.isDartCoreBool ||
+        iterableGenericType.isDartCoreDouble ||
+        iterableGenericType.isDartCoreInt ||
+        iterableGenericType.isDartCoreNum ||
+        iterableGenericType.isDartCoreString;
+    //
```

Comment:
delete this

Reply:
If you delete this it will not cover current null-safe use cases.  The original code did not provide good coverage for existing test cases.  The current test cases are good for new code as well.

Reply:
done